### PR TITLE
dialects: (ptr) Introducing ptr dialect

### PR DESCRIPTION
**Pull request by mamanain**
Porting the ptr dialect proposed [here](https://discourse.llvm.org/t/rfc-ptr-dialect-modularizing-ptr-ops-in-the-llvm-dialect/75142/21). Going to transfer the riscv compilation pipeline to it.

### CODE REVIEW

File: xdsl/dialects/__init__.py [hunk 1]
```diff
@@ -360,6 +365,7 @@ def get_transform():
         "onnx": get_onnx,
         "pdl": get_pdl,
         "printf": get_printf,
+        "opaque_ptr": get_opaque_ptr,
```
Comment (superlopuh) [inline xdsl/dialects/__init__.py:368]: ? why not ptr? is it to avoid the confusion with MLIR for unknown operations in dialects?

Comment (mamanain) [reply xdsl/dialects/__init__.py:368]: yep, exactly

Comment (superlopuh) [reply xdsl/dialects/__init__.py:368]: Hmm, I feel like we've had a discussion about this in the past, and I don't remember the conclusion. I would recommend naming this `xdsl_ptr` or `ptr_xdsl`, to make it clear what's happening.

File: tests/filecheck/dialects/ptr/ops.mlir [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// RUN: xdsl-opt %s --parsing-diagnostics --verify-diagnostics --split-input-file | filecheck %s
+
+builtin.module {
```
Comment (superlopuh) [inline tests/filecheck/dialects/ptr/ops.mlir:3]:
```suggestion
```

File: xdsl/dialects/ptr.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+class PtrType(ParametrizedAttribute, TypeAttribute):
+    name = "opaque_ptr.ptr"
+
+
+@irdl_op_definition
+class PtrAddOp(IRDLOperation):
+    name = "opaque_ptr.ptradd"
+
+    addr = operand_def(PtrType)
+    offset = operand_def(AnyOf([IntegerType, IndexType]))
```
Comment (superlopuh) [inline xdsl/dialects/ptr.py:37]:
```suggestion
    offset = operand_def(IntegerAttrTypeConstr)
```

File: xdsl/dialects/ptr.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    assembly_format = "$addr `,` $offset attr-dict `:` `(` type($addr) `,` type($offset) `)` `->` type($result)"
+
+
+# haven't managed to pass a type here yet. so did it with a hack.
+@irdl_op_definition
+class TypeOffsetOp(IRDLOperation):
+    name = "opaque_ptr.type_offset"
+
+    elem_type = prop_def(base(Attribute))
+    offset = result_def(AnyOf([IntegerType, IndexType]))
```
Comment (superlopuh) [inline xdsl/dialects/ptr.py:49]:
```suggestion
    elem_type = prop_def()
    offset = result_def(IntegerAttrTypeConstr)
```

File: xdsl/dialects/ptr.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    name = "opaque_ptr.to_ptr"
+
+    source = operand_def(MemRefType)
+    res = result_def(PtrType)
+
+    assembly_format = "$source attr-dict `:` type($source) `->` type($res)"
+
+
+Ptr = Dialect(
+    "opaque_ptr", [PtrAddOp, TypeOffsetOp, StoreOp, LoadOp, ToPtrOp], [PtrType]
```
Comment (superlopuh) [inline xdsl/dialects/ptr.py:92]:
```suggestion
    "opaque_ptr", [PtrAddOp, TypeOffsetOp, StoreOp, LoadOp, ToPtrOp,], [PtrType,]
```

File: tests/filecheck/dialects/ptr/ops.mlir [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  %p, %idx, %v, %m = "test.op"() : () -> (!opaque_ptr.ptr, index, i32, memref<10xi32>)
+
+  // CHECK: %r0 = opaque_ptr.ptradd %p, %idx : (!opaque_ptr.ptr, index) -> !opaque_ptr.ptr
+  %r0 = opaque_ptr.ptradd %p, %idx : (!opaque_ptr.ptr, index) -> !opaque_ptr.ptr
+
+  // CHECK-NEXT: %r1 = "opaque_ptr.type_offset"() <{"elem_type" = i32}> : () -> index
+  %r1 = "opaque_ptr.type_offset"() <{"elem_type" = i32}> : () -> index
```
Comment (superlopuh) [inline tests/filecheck/dialects/ptr/ops.mlir:10]: let's add a custom syntax here? happy to go with what looks prettiest to you

File: xdsl/dialects/ptr.py (hunk 4 above)
Comment (mamanain) [reply xdsl/dialects/ptr.py:49]: `prop_def` complains that it needs a constraint parameter

File: xdsl/dialects/ptr.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    assembly_format = "$addr `,` $offset attr-dict `:` `(` type($addr) `,` type($offset) `)` `->` type($result)"
+
+
+# haven't managed to pass a type here yet. so did it with a hack.
+@irdl_op_definition
+class TypeOffsetOp(IRDLOperation):
+    name = "ptr_xdsl.type_offset"
+
+    elem_type = prop_def(base(Attribute))
```
Comment (superlopuh) [inline xdsl/dialects/ptr.py:47]:
```suggestion
    elem_type = prop_def(Attribute)
```
You're right, I didn't realise that `prop_def` needed a constraint. Also `AnyAttr()` would work here.

**Review by superlopuh — APPROVED** [review]
LGTM modulo `base(Attribute)`

**Pull request merged**